framework/include/base/MooseApp.h

```diff
@@ -99,6 +99,17 @@ class MooseApp : public ConsoleStreamInterface, public libMesh::ParallelObject
   virtual void run();
 
   /**
+   * Returns the version (revision string) of the application. Applications may have several sub
+   * apps so a virtual is provided for user-defined version strings. By default, the version of the
+   * framework is printed. The framework, test, modules, examples, tutorials all share the same
+   * version.
+   */
+  virtual std::string getRevisionString() const;
+
+  /// Returns the current version of the framework
+  std::string getVersion() const;
```

Comment:
Why is revision virtual and version is not?

Also the comment style should be `/**...*/`

Reply:
Good question. I'm not sure I want people to override this. If we plan to try to parse it, it'll need to be consistent. I can fix the comment. Do you think we need them both to be virtual?

Reply:
So let's think about how this will behave in an application. What we do in RELAP is that when people do `./relap-7-opt --version` they get RELAP version not framework version. Since this is in `MooseApp`, I'd assume that `getVersion()` is going to give me my application version, not MOOSE version. But then, this has to be virtual and `getVersion()` has to return `MYAPP_VERSION` from the generated header file.

Reply:
In addition to this, we might need a function that gives back MOOSE version/revision.

Reply:
Also, if I run `--version` in an application I do not want to see `Framework revision: <APPLICATION_REVISION>, Version: <APPLICATION_VERSION>`after I override `getVersion`/`getRevision`. I would rather see something `My application name, version <APPLICATION_VERSION>`, i.e. your patch does not give that flexibility.